sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/JavaUdfLoader.java

```diff
@@ -0,0 +1,217 @@
+      if (!functionDefinitions.javaScalarFunctions.containsKey(functionPath)) {
+        throw new IllegalArgumentException(
+            String.format(
+                "No implementation of scalar function %s found in %s.\n"
+                    + " 1. Create a class implementing %s and annotate it with @AutoService(%s.class).\n"
+                    + " 2. Add function %s to the class's userDefinedScalarFunctions implementation.",
+                functionFullName,
+                jarPath,
+                UdfProvider.class.getSimpleName(),
+                UdfProvider.class.getSimpleName(),
+                functionFullName));
+      }
+      return functionDefinitions.javaScalarFunctions.get(functionPath);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format(
+              "Failed to load user-defined scalar function %s from %s", functionFullName, jarPath),
+          e);
+    }
+  }
+
+  /**
+   * Load a user-defined aggregate function from the specified jar.
+   *
+   * <p><strong>WARNING</strong>: The first time a jar is loaded, it is added to the thread's
+   * context {@link ClassLoader} so that the jar can be staged by the runner.
+   */
+  public Combine.CombineFn loadAggregateFunction(List<String> functionPath, String jarPath) {
+    String functionFullName = String.join(".", functionPath);
+    try {
+      UserFunctionDefinitions functionDefinitions = loadJar(jarPath);
+      if (!functionDefinitions.javaAggregateFunctions.containsKey(functionPath)) {
+        throw new IllegalArgumentException(
+            String.format(
+                "No implementation of aggregate function %s found in %s.\n"
+                    + " 1. Create a class implementing %s and annotate it with @AutoService(%s.class).\n"
+                    + " 2. Add function %s to the class's userDefinedAggregateFunctions implementation.",
+                functionFullName,
+                jarPath,
+                UdfProvider.class.getSimpleName(),
+                UdfProvider.class.getSimpleName(),
+                functionFullName));
+      }
+      return functionDefinitions.javaAggregateFunctions.get(functionPath);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format(
+              "Failed to load user-defined aggregate function %s from %s",
+              functionFullName, jarPath),
+          e);
+    }
+  }
+
+  private ClassLoader createAndSetClassLoader(String inputJarPath) throws IOException {
+    Preconditions.checkArgument(!inputJarPath.isEmpty(), "Jar path cannot be empty.");
+    ResourceId inputJar = FileSystems.matchNewResource(inputJarPath, false /* is directory */);
+    File tmpJar = File.createTempFile("sql-udf-", inputJar.getFilename());
+    FileSystems.copy(
+        Collections.singletonList(inputJar),
+        Collections.singletonList(
+            FileSystems.matchNewResource(tmpJar.getAbsolutePath(), false /* is directory */)));
```

Comment:
Is it possible to have a verification after the copy is done? E.g. verify file size or MD5 hash? I am thinking if user query has an `function not found` issue due to a bad downloaded jar, we can locate the root cause based on log? Otherwise such issue might be hard to debug.

Reply:
I logged the MD5 hash.